packages/signing-utils/src/ssh-client.ts

```diff
@@ -0,0 +1,106 @@
+import type { ConnectConfig, SFTPWrapper } from 'ssh2';
+import { Client } from 'ssh2';
+import { readFile } from 'fs/promises';
+import { debug } from './utils';
+
+export class SSHClient {
+  private sshConnection: Client;
+  private sftpConnection?: SFTPWrapper;
+
+  private connected = false;
+
+  constructor(private sshClientOptions: ConnectConfig) {
+    this.sshConnection = new Client();
+    this.setupEventListeners();
+  }
+
+  setupEventListeners() {
+    this.sshConnection.on('ready', () => {
+      debug('SSH: Connection established');
+      this.connected = true;
+    });
+    this.sshConnection.on('error', (err) => {
+      debug('SSH: Connection error', err);
+      this.connected = false;
+    });
+    this.sshConnection.on('close', () => {
+      debug('SSH: Connection closed');
+      this.connected = false;
+      this.sshConnection.destroy();
+    });
+  }
+
+  async connect() {
+    if (this.connected) {
+      return Promise.resolve();
+    }
+    const privateKey = this.sshClientOptions.privateKey
+      ? await readFile(this.sshClientOptions.privateKey)
+      : undefined;
+    return new Promise((resolve, reject) => {
+      this.sshConnection.connect({
+        ...this.sshClientOptions,
+        privateKey,
+      });
+      this.sshConnection.on('error', reject);
+      // @ts-expect-error We expect an error here - why?
+      this.sshConnection.on('ready', resolve);
+    });
+  }
+
+  disconnect() {
+    this.sshConnection.end();
+    this.connected = false;
+  }
+
+  async exec(command: string): Promise<string> {
+    if (!this.connected) {
+      throw new Error('Not connected to ssh server');
+    }
+    return new Promise((resolve, reject) => {
+      this.sshConnection.exec(command, (err, stream) => {
+        if (err) {
+          return reject(err);
+        }
+        let data = '';
+        stream.on('data', (chunk: string) => {
```

Comment:
```suggestion
        stream.setEncoding('utf8').on('data', (chunk: string) => {
```

(otherwise you'll into problems with multibyte characters split across chunks and the parameter type signature would be `chunk: Buffer`, not `chunk: string`)

Reply:
This is interesting, not something I've ever considered. Thanks for sharing!

I made the change, suggested here, but just because I'm curious – would leaving the encoding unspecified, concatenating buffers in each callback, and then stringifying the result at the end achieve the same result?  I'm not sure if that would be more performance, because I don't know of a way to concatenate buffers without copying the original. In the driver, we maintain a custom data structure for this purpose.

Reply:
Yes, gathering all chunks as Buffers and then concatenating them would give the same result – I'd expect it to be significantly slower, though, there's a lot more objects and crossings between JS and C++ involved in that case :)